### ISSUE TITLE
Improve/fix UX when integrating walletconnect-v2 + metamask-mobile in a multichain dapp

### ISSUE THREAD

**Issue by paro-paro** [MetaMask/metamask-mobile#7138]
I have found the following issue that affects the user experience when integrating walletconnect-v2 + metamask-mobile in a multichain dapp. 

You can find a reproduction repo [here](https://github.com/paro-paro/poc-wc-mm) in case is useful.

---

Steps to reproduce:

1) Init your ethereum provider with two networks (ETH and CELO)

```
const provider = await EthereumProvider.init({
  projectId: 'foo',
  chains: [1],
  optionalChains: [42220], // celo
  showQrModal: true,
})
```

2) Connect you metamask mobile wallet by scanning the QR code (make sure metamask is connected to Ethereum Mainnet). 

3) Switch to CELO network from the dapp with the `wallet_switchEthereumChain` method. 

```
await provider.request({
   method: 'wallet_switchEthereumChain',
      params: [
        {
          chainId: '0xa4ec',
        },
      ],
    })
```

4) Trigger a `personal_sign` request.

So far, so good. 

Metamask will connect successfully to the dapp (2), the `wallet_switchEthereumChain` will prompt the user to change the network to CELO (3) and the `personal_sign` prompt will also be displayed in the metamask wallet (4).

But, once in this state, if the user...

5) Switches back to ETH network.

In this case, metamask will not prompt the user to change the network but the provider will effectively reflect this change on the `chainChanged` event. I believe this is expected behaviour (changing again to CELO will behave the same way, not prompting the user but reflecting the change in the provider).

6) Triggers again the `personal_sign` request.

Here is when I find the problem: 

Metamask will properly detect that the provider is connected to ETH and before displaying the `personal_sign` prompt it will ask the user to change the network first (since metamask is still connected to CELO). But after that, it wont sequentially display the sign prompt. The user will have to trigger again the `personal_sign` request (by, for example, clicking the sign message button twice which is unintuitive and feels like the dapp is not working). 

Fortunately, this is the expected behaviour we get if the user switches again to the CELO network and launches the sign request: metamask will prompt the user for a network change (really nice) and right after that will prompt the sign request (so perfect). 

I will expect metamask to behave like this always, making the UX in multichain dapps quite good. 

--- 

Thanks in advanced!

**Comment by andreahaku** [comment]
@paro-paro we are aware of this kind of UX issue due to the way the MM mobile app handles this kind of sequences of RPC calls requiring modals to be displayed. Thanks for reporting.

**Comment by christopherferreira9** [comment]
Hi @paro-paro ! As of the latest version of the wallet `7.14.0` it seems that this behaviour has been fixed. Changing the chain is being synced with the dapp but in the following video I forced them to not be synced by requesting the sign as fast as possible before the event would throw. In this case as you stated the wallet requests to change chain but this time around it also presents the sign request right after switching chains. 

As for the scope of this issue it is working as intended now but the UX flaw present in https://github.com/MetaMask/metamask-mobile/issues/7878 still persists. I'm closing this its favour.

https://github.com/MetaMask/metamask-mobile/assets/104831203/f26bdbaf-ea08-40ea-a40d-712c1299d7cf
